osquery/core/database/rocksdb_database.cpp

```diff
@@ -0,0 +1,388 @@
+  options.OptimizeForSmallDb();
+
+  // Set meta-data (mostly) handling options.
+  options.create_if_missing = true;
+  options.create_missing_column_families = true;
+  options.error_if_exists = false;
+  options.paranoid_checks = true;
+
+  // Please review RocksDBLogger if you want to modify this line!
+  options.info_log =
+      std::make_shared<RocksDBLogger>(rocksdb::InfoLogLevel::ERROR_LEVEL);
+  options.info_log_level = rocksdb::ERROR_LEVEL;
+
+  options.max_open_files = 128;
+
+  options.log_file_time_to_roll = 0;
+  options.keep_log_file_num = 10;
+  options.max_log_file_size = 1024 * 1024 * 1;
+
+  options.stats_dump_period_sec = 0;
+  options.max_manifest_file_size = 1024 * 500;
+
+  // Performance and optimization settings.
+  options.compression = rocksdb::kNoCompression;
+  options.compaction_style = rocksdb::kCompactionStyleLevel;
+  options.arena_block_size = (4 * 1024);
+  options.write_buffer_size = (4 * 1024) * FLAGS_rocksdb_buffer_blocks;
+  options.max_write_buffer_number =
+      static_cast<int>(FLAGS_rocksdb_write_buffer);
+  options.min_write_buffer_number_to_merge =
+      static_cast<int>(FLAGS_rocksdb_merge_number);
+
+  return options;
+}
+
+std::vector<rocksdb::ColumnFamilyDescriptor>
+RocksdbDatabase::createDefaultColumnFamilies(const rocksdb::Options& options) {
+  std::vector<rocksdb::ColumnFamilyDescriptor> column_families;
+  column_families.push_back(rocksdb::ColumnFamilyDescriptor(
+      rocksdb::kDefaultColumnFamilyName, options));
+  for (const auto& cf_name : kDomains) {
+    column_families.push_back(
+        rocksdb::ColumnFamilyDescriptor(cf_name, options));
+  }
+  return column_families;
+}
+
+void RocksdbDatabase::close() {
+  VLOG(1) << "Closing db... ";
+  assert(db_ != nullptr && "db connection is not open");
+  if (db_ != nullptr) {
+    handles_map_.clear();
+    db_ = nullptr;
+  }
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::destroyDB() {
+  VLOG(1) << "Destroying db at path " << path_;
+
+  assert(db_ == nullptr && "can't destroy db while it in use");
+  if (db_ != nullptr) {
+    // Try to recover in case if db was open
+    close();
+  }
+
+  rocksdb::Options options = getOptions();
+  auto db_path = boost::filesystem::path(path_).make_preferred();
+  auto destroy_status = rocksdb::DestroyDB(db_path.string(), options);
+  if (!destroy_status.ok()) {
+    return createError(DatabaseError::FailToDestroyDB, "Fail to destroy db: ")
+           << destroy_status.ToString();
+  }
+  return Success();
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::openInternal(
+    const rocksdb::Options& options, const boost::filesystem::path& path) {
+  VLOG(1) << "Will try to open db at path: " << path.string();
+  auto column_families = createDefaultColumnFamilies(options);
+  auto db_path = boost::filesystem::path(path).make_preferred();
+  auto db_parent_path_status = boost::filesystem::status(db_path.parent_path());
+  if (!boost::filesystem::exists(db_parent_path_status) ||
```

Comment:
Can we store this variables? Error message will be easier to understand and also there will be no chance of this variables being changed before constructing the message